File: ADT/trees/623.js
var addOneRow = function (root, val, depth) {
    if (depth === 1) {
        const newRoot = new TreeNode(val);
        newRoot.left = root;
        return newRoot;
    }

    const add = (node, currentDepth) => {
        if (!node) return null;

        if (currentDepth === depth - 1) {
            const oldLeft = node.left;
            const oldRight = node.right;

            node.left = new TreeNode(val);
            node.right = new TreeNode(val);

            node.left.left = oldLeft;
            node.right.right = oldRight;

            return;
        }

        add(node.left, currentDepth + 1);
        add(node.right, currentDepth + 1);
    };

    add(root, 1);
    return root;
};

File: ADT/trees/623.py
class Solution:
    def add(
        self, root: Optional[TreeNode], val: int, depth: int, curr: int
    ) -> Optional[TreeNode]:
        if not root:
            return None

        if curr == depth - 1:
            lTemp = root.left
            rTemp = root.right

            root.left = TreeNode(val)
            root.right = TreeNode(val)
            root.left.left = lTemp
            root.right.right = rTemp

            return root

        root.left = self.add(root.left, val, depth, curr + 1)
        root.right = self.add(root.right, val, depth, curr + 1)

        return root

    def addOneRow(
        self, root: Optional[TreeNode], val: int, depth: int
    ) -> Optional[TreeNode]:
        if depth == 1:
            newRoot = TreeNode(val)
            newRoot.left = root
            return newRoot

        return self.add(root, val, depth, 1)
